Cellular ga

nasseralbunian opened the pull request.

jose commented:
Hi @nasseralbunian,

Thank you for your pull request, I'll have a look as soon as possible.

jose commented:
@nasseralbunian, before I go through all your changes, here a few comments:

- Class ChromosomeCollection is never used, can we remove it?
- How about we move Positions enum from Properties to Neighbourhood or NeighborModels class?
- In Properties, I would be in favour of renaming CGA to a more descriptive name, e.g., CellularGA. What do you think?
- I failed to see why do we need another `keeptheOffspring` function. Can you please explain why? It looks like `keeptheOffspring` does the same as the existing (deprecated) method `keepOffspring`, but I might be wrong...
- Could please address the warnings of CellularGA, and Neighbourhood classes? And remove the code that is not used or that is commented out?

PS: Can you please point me the paper that describes this algorithm?


--
Thanks,
Jose

nasseralbunian commented:
- Class `ChromosomeCollection` is never used, can we remove it?
Yes, it should be removed.

- How about we move `Positions` enum from `Properties` to `Neighbourhood` or `NeighborModels` class?
We can do that.

- In Properties, I would be in favour of renaming CGA to a more descriptive name, e.g., CellularGA. What do you think?
Great idea.

- I failed to see why do we need another `keeptheOffspring` function. Can you please explain why? It looks like `keeptheOffspring` does the same as the existing (deprecated) method `keepOffspring`, but I might be wrong...
They are the same. I decided to declare a new function (`keeptheOffspring`) since `keepOffspring` is deprecated. At the end, the purpose is to compare a parent to an offspring.

- Could please address the warnings of CellularGA, and Neighbourhood classes? And remove the code that is not used or that is commented out?
Sure. I will do.

- Can you please point me the paper that describes this algorithm?
Well, the main reference that I considered is [this book](https://books.google.co.uk/books?id=xcvsd6FmhyMC&lpg=PR1&dq=cellular%20genetic%20algorithm&pg=PR1#v=onepage&q=cellular%20genetic%20algorithm&f=false). The algorithm is explained clearly in this book.

jose commented:
Hi @nasseralbunian, have you done all those points you listed?

nasseralbunian commented:
Hi @jose, yes

jose commented:
Well, you last commit [e55809f](https://github.com/EvoSuite/evosuite/pull/141/commits/e55809f28cd81fa64da86c65a691adc9996f4ec8) only addressed warnings and removed the code that was comment out.

- class `ChromosomeCollection` is still there.
- `Positions` enum still lives in `Properties` class
- CellularGA is still called CGA
- do we need `keeptheOffspring` or not?

nasseralbunian commented:
Hi @jose 

- class `ChromosomeCollection` is still there.
Deleted.

- `Positions` enum still lives in `Properties` class
I moved them to the `NeighborModels` class and removed them from the `Properties` class.

- `CellularGA` is still called `CGA`
I changed the class name and also the `Algorithm` property in `Properties` class.

- Do we need `keeptheOffspring` or not?
Deleted and replaced by `keepOffspring`

jose commented:
Thanks @nasseralbunian.

- Why did you remove the annotation `Deprecated` of `keepOffspring` method? I would keep it as deprecated, because (and as the javadoc says) "should not be used, as it does not handle `Properties.CHECK_PARENTS_LENGTH`".
- Can you please give more descriptive names to `CGA_Models`, e.g., what does `COMNIN` mean?
- Can you please add unit tests to those methods defined in Neighbourhood class? (see for example the unit tests I created for some functions of the SPEA2 algorithm in org.evosuite.ga.metaheuristics.SPEA2Test -- master module)
- Can you please add (at least one) system test to CellularGA? (you can find an example of a system test in org.evosuite.ga.metaheuristics.SPEA2SystemTest -- master module)

nasseralbunian commented:
Hi @jose 

The unit tests that I added are to test how the neighbors of an individual are selected. For the system test, I need more time as I still find it difficult to create one.

jose commented on client/src/main/java/org/evosuite/ga/ReplacementFunction.java:
```diff
@@ -161,4 +161,17 @@ protected int compareBestOffspringToBestParent(Chromosome parent1,
+	 *            a {@link org.evosuite.ga.Chromosome} object.
+	 * @return a int.
+	 */
+	protected int compareOffspringToParent(Chromosome chromosome1, Chromosome chromosome2) {
```
This function does not seem to be used anywhere. Should we remove it?

jose commented on client/src/main/java/org/evosuite/ga/metaheuristics/CellularGA.java:
```diff
@@ -0,0 +1,369 @@
+	protected ReplacementFunction replacementFunction;
+	
+	/** Size of population **/
+	private int populationSize=0;
```
We do not need yet another variable, you can always use Properties.POPULATION.

jose commented on client/src/main/java/org/evosuite/ga/metaheuristics/CellularGA.java:
```diff
@@ -0,0 +1,369 @@
+			}
+			
+			// replace-if-better policy
+			if(keepOffspring(mainIndividual, tempIndividual)){
```
How about we use the isBetterOrEqual method from GeneticAlgorithm class?

jose commented on client/src/main/java/org/evosuite/ga/metaheuristics/CellularGA.java:
```diff
@@ -0,0 +1,369 @@
+			notifyEvaluation(offspring2);
+		}
+		
+		if(keepOffspring(offspring1, offspring2))
```
How about we use the isBetterOrEqual method from GeneticAlgorithm class?

jose commented on client/src/main/java/org/evosuite/ga/metaheuristics/CellularGA.java:
```diff
@@ -0,0 +1,369 @@
+	 * @return a boolean
+	 */
+	@SuppressWarnings("deprecation")
+	protected boolean keepOffspring(Chromosome offspring1, Chromosome offspring2) {
```
If we use the isBetterOrEqual method from GeneticAlgorithm class, we could remove this method.

jose commented on master/src/test/java/org/evosuite/ga/metaheuristics/CellularGASystemTest.java:
```diff
@@ -0,0 +1,73 @@
+	    Properties.STOPPING_CONDITION = sc;
+	    Properties.SEARCH_BUDGET = budget;
+	    Properties.MINIMIZE = false;
+	    Randomness.setSeed(10);
```
You don't need to set the _seed_, SystemTestBase class already does that for you.

jose commented:
@nasseralbunian just to let you know I've just merged this pull request. Thank you.
